samcli/commands/build/build_context.py

```diff
@@ -0,0 +1,94 @@
+
+import os
+import shutil
+import yaml
+
+from samcli.yamlhelper import yaml_parse
+from samcli.commands.local.lib.sam_function_provider import SamFunctionProvider
+
+
+class BuildContext(object):
+
+    def __init__(self, template_file, source_root, build_dir, clean=False, use_container=False):
+        self._template_file = template_file
+        self._source_root = source_root
+        self._build_dir = build_dir
+        self._clean = clean
+        self._use_container = use_container
+
+        self._function_provider = None
+        self._template_dict = None
+        self._app_builder = None
+
+    def __enter__(self):
+        self._template_dict = self._get_template_data(self._template_file)
+        self._function_provider = SamFunctionProvider(self._template_dict)
+
+        self._build_dir = self._setup_build_dir()
+
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @property
+    def function_provider(self):
+        return self._function_provider
```

Comment:
Should we go one level further and just provide a class that conforms to a provider instead of a specific one?

Reply:
Possibly. But this is a good easy place to start